hibernate-core/src/main/java/org/hibernate/type/descriptor/java/LocaleTypeDescriptor.java

```diff
@@ -51,43 +51,33 @@ public Locale fromString(String string) {
 			return Locale.ROOT;
 		}
 
-		boolean separatorFound = false;
-		int position = 0;
-		char[] chars = string.toCharArray();
-
-		for ( int i = 0; i < chars.length; i++ ) {
-			// We just look for separators
-			if ( chars[i] == '_' ) {
-				if ( !separatorFound ) {
-					// On the first separator we know that we have at least a language
-					string = new String( chars, position, i - position );
-					position = i + 1;
-				}
-				else {
-					// On the second separator we have to check whether there are more chars available for variant
-					if ( chars.length > i + 1 ) {
-						// There is a variant so add it to the constructor
-						return new Locale( string, new String( chars, position, i - position ), new String( chars,
-								i + 1, chars.length - i - 1 ) );
+		Locale.Builder builder = new Locale.Builder();
+		String[] parts = string.split("_");
+
+		for (int i = 0; i < parts.length; i++) {
+			String s = parts[i];
+			switch (i) {
+				case 0:
+					builder.setLanguage(s);
+					break;
+				case 1:
+					builder.setRegion(s);
+					break;
+				case 2:
+					if (i < parts.length - 1 || !s.startsWith("#")) {
+						builder.setVariant(s);
+						break;
 					}
-					else {
-						// No variant given, we just have language and country
-						return new Locale( string, new String( chars, position, i - position ), "" );
+				case 3:
+					if (s.startsWith("#")) {
+						s = s.substring(1);
```

Comment:
It would be ideal to add `break` for `case 3` for best practice. Imagine in the future a new case is needed and it might create a subtle bug that the coder forgets to insert `break` for `case 3`.

Reply:
The gist of this PR seems to lie in the above `case 2` and `case 3`. There is subtlety here that if s starts with `#` in `case 2`, it is equivalent to `case 3`. It would be great to add some comment like `// intentionally fall through` before `case 3` to help future code maintainer to understand the subtlety here.

Reply:
TBH, there must be still room for code readability improvement by clearer logic. The break case falling through is to difficult to understand.

Reply:
I agree that the above code is less than ideal. Adding a break to case 2 would mean duplicating the code from case 3, I don't know which one is the lesser of evils. We could perhaps just avoid the switch altogether and unroll it into 4 if statements.

It could perhaps be rewritten to mimic this pattern as seen in [LanguageTag](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/share/classes/sun/util/locale/LanguageTag.java#L202), which would require using something like StringTokenizer. It is unfortunate that this code cannot simply use [toLanguageTag](https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/util/Locale.html#toLanguageTag()) / [forLanguageTag](https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/util/Locale.html#forLanguageTag(java.lang.String))...

Reply:
IMHO, using StringTokenizer is perfectly fine in Hibernate core codebase (it has been used more than 50 after a quick search). Code readability is a much bigger concern. The current switch case falling through is a clever approach but might make future code maintaining much difficult.